### ISSUE TITLE
[RFC] extend the virtualenv generator to expose the CONAN include path

### ISSUE THREAD

**Issue by cinghiale** [conan-io/conan#441]
### The idea:

What do you think to expose the conan include path in the shell where the virtualenv is activated?

To be clear I want to add a line like this one to [virtualenv.py](https://github.com/conan-io/conan/blob/develop/conans/client/generators/virtualenv.py)

``` python
activate_lines.append("export _CONAN_INCLUDE_PATH=%s" % ':'.join(self.build_info.include_paths + self.deps_build_info.include_paths))
```
### Why I need it:

With this patch I can instruct vim to append `_CONAN_INCLUDE_PATH` to one of the environment variables read by `gcc` so that my linter plugin `syntastic` can found the libraries installed by conan.

I am deliberatly not proposing to overwrite one of the standard env variable (like `CPLUS_INCLUDE_PATH`) to not interfer with the normal workflow

If you like the idea I can prepare a PR

**Comment by memsharded** [comment]
Might be interesting, but is it possible to directly generate something that ``syntastic`` would understand? We could do like the YouCompleteMe generator for vim, maybe that would be even more convenient.

Does it work only in *nix systems?, because ``SET`` would be necessary in windows instead of ``export``

I think it can be useful and doable, but maybe the feedback of @lasote is necessary here. Thanks very much for the proposal

**Comment by cinghiale** [comment]
Obviously the PR must include the relative changes to support Windows. Probably the real issue will be choose a relevant name for the env variable

**Comment by lasote** [comment]
I don't know very much about `syntastic` but maybe we should keep the virtualenv generator to set the declared env_info variables in the conanfiles.
I think the right way is the @memsharded proposition about creating a new generator for `syntactic`. 

The FAQ *4.15. My favourite checker needs to load a configuration file from the project's root rather than the current directory* sounds like maybe it's possible. Am I right? 
@cinghiale, Can `syntactic` load the needed information from a file? If true, how does it look like? If you write us an example I think we can easily develop a new generator.

Thanks!

**Comment by cinghiale** [comment]
@lasote Syntastic is a plugin that run a different checker according to the current file type.
It doesn't have a configuration file besides `~/.vimrc`.

Exposing the include and lib path in a `virtualenv` is also useful to write a quick test program; when you slap some code in a `foo.cpp` directly in the `build` dir and run g++ to just test some corner case.

**Comment by lasote** [comment]
@cinghiale but the quick test program can use the gcc generator. And I suppose, for the same reason, we should expose the compiler flags, linker flags... etc... I think it can be a lot of noise but I'm going to play with syntastic and think about it.

**Comment by lasote** [comment]
Hi! @cinghiale syntastic can read a **".syntastic_cpp_config"** file in your project. The content could be just:

```-I /home/bla/bla/gtest/1.7.0/lasote/stable/package/41b4a6b644198013c12a53602fa3d/include```

So I think it's the best way is program a simple generator that creates the ".syntastic_cpp_config" file executing the "conan install" command.

If you want to try it you can build a custom generator for syntastic using the conan docs:
http://conanio.readthedocs.io/en/latest/howtos/dyn_generators.html?highlight=generators

The generator only needs to add the include paths to the file.
We can help you if you need it!

**Comment by cinghiale** [comment]
I can write this that's not a problem.

There is the annoyance that conan is run in the build directory while vim works on the project dir

I still can find some advantages in a generic solution like the one I proposed, what are the cons to expose the include dirs in the virtualenv?

**Comment by lasote** [comment]
You can run conan in a different directory and generate the file where you want:

```
mkdir build
cd build
conan_src install --file ../conanfile.py
```
It could help with the annoyance?

I don't want to discard the generic solution but I think is more straight forward to generate a file and run vim without the step of teaching vim how to do with the environment variable. 

If you find another real use case for export those paths in the virtual environment I will be glad to add it but for syntastic I think it's better to make a simple generator.

**Comment by cinghiale** [comment]
I'm ok with your decision; from my POV (linux+cmdline guy) a generic composable solution has more value than an specific generator for a niche (?) use cases but, as I said, it's my pov :)

BTW I don't understand your snipped:

```
cd project # here there is conanfile.txt and here I want .syntastic_cpp_config
mkdir build
cd build
conan install .. # .syntastic_cpp_config will be generated here 
```

**Comment by lasote** [comment]
I meant that you can generate the .syntastic_cpp_config file in the folder that you want, not necessarily the same where you have your conanfile.py. The snippet just shows this feature.

**Comment by memsharded** [comment]
Actually, I have another issue #408, because it would be useful to add to LD_LIBRARY_PATH the conan packages paths with shared libraries, so they can be found to run programs while working in conan. I think such feature could also possibly export other environment variables as the include paths, library paths or library names, I see not special harm in it. The main thing is that the virtualenvs composability should be studied and checked, as the main use case is to install dev-tools (as cmake, mingw), but that is done with a different recipe, different from the project requirements. So two virtualenvs should be loaded and nested, and that is still not sufficiently tested.

So I think they are two different issues, for syntastic, creating the files that syntastic uses to define its input, is the way to go, what syntastic users would expect and users don't need to manually handle the translation from env-variable to syntastic.   

The other issue would be augment #408 with more information for packages, like include dirs, etc. to achieve the composability, convenience for cmdline, testing, etc.

**Comment by memsharded** [comment]
Hi @cinghiale ,

I think this idea has been addressed by conan 0.20 new ``virtualbuildenv`` which is composable to the ``virtualenv``. Please check http://conanio.readthedocs.io/en/latest/integrations/virtualenv.html#virtualbuildenv-environment, and tell us otherwise. Thanks!
